File: python/stack_and_queue/array_stack.py
"""
基于数组实现的栈，将数组的尾部视为栈顶，入栈和出栈都操作尾部
"""


class ArrayStack:
    def __init__(self):
        self._stack: list[int] = []

    def size(self):
        return len(self._stack)

    def is_empty(self):
        return self._stack == []

    def push(self, value: int):
        self._stack.append(value)

    def pop(self) -> int:
        self.check_index()
        return self._stack.pop()

    def peek(self) -> int:
        self.check_index()
        return self._stack[-1]

    def to_list(self) -> list:
        return self._stack

    def check_index(self) -> None:
        if self.is_empty():
            raise IndexError("栈为空")


def main():
    array_stack = ArrayStack()
    assert array_stack.is_empty() is True

    # array_stack.peek()  # index error

    array_stack.push(1)
    assert 1 == array_stack.peek()
    assert 1 == array_stack.pop()
    array_stack.push(1)
    array_stack.push(3)
    array_stack.push(2)
    array_stack.push(5)
    array_stack.push(4)
    assert [1, 3, 2, 5, 4] == array_stack.to_list(), array_stack.to_list()


if __name__ == "__main__":
    main()

File: go/stack_and_queue/array_stack.go
// Package stack_and_queue  基于数组的栈，利用 slice 实现
package stack_and_queue

type ArrayStack struct {
	data []int
}

func NewArrayStack() *ArrayStack {
	return &ArrayStack{
		data: []int{},
	}
}

func (s *ArrayStack) Push(value int) {
	s.data = append(s.data, value)
}

func (s *ArrayStack) Pop() any {
	v := s.Peek()
	s.data = s.data[:s.Size()-1]
	return v
}

func (s *ArrayStack) Peek() any {
	if s.IsEmpty() {
		return nil
	}
	return s.data[s.Size()-1]
}

func (s *ArrayStack) Size() int {
	return len(s.data)
}

func (s *ArrayStack) IsEmpty() bool {
	return s.Size() == 0
}

func (s *ArrayStack) ToSlice() []int {
	return s.data
}
